### PR TITLE
[AArch64] Add assembly/disassembly for FMOP4A (widening, 2-way, FP8 to FP16) instructions

### PR DESCRIPTION
**Pull request by momchil-velikov**
The new instructions are described in
https://developer.arm.com/documentation/ddi0602/2024-09/SME-Instructions

### CODE REVIEW

File: llvm/test/MC/AArch64/SME2p2/fmop4a-fp8-fp16-widening.s [hunk 1]
```diff
@@ -0,0 +1,95 @@
+// RUN: llvm-mc -triple=aarch64 -show-encoding -mattr=+sme2p2,+sme-f8f16 < %s \
+// RUN:        | FileCheck %s --check-prefixes=CHECK-ENCODING,CHECK-INST
+// RUN: not llvm-mc -triple=aarch64 -show-encoding < %s 2>&1 \
+// RUN:        | FileCheck %s --check-prefix=CHECK-ERROR
+// RUN: llvm-mc -triple=aarch64 -filetype=obj -mattr=+sme2p2,+sme-f8f16 < %s \
+// RUN:        | llvm-objdump -d --mattr=+sme2p2,+sme-f8f16 - | FileCheck %s --check-prefix=CHECK-INST
+// RUN: llvm-mc -triple=aarch64 -filetype=obj -mattr=+sme2p2,+sme-f8f16 < %s \
+// RUN:        | llvm-objdump -d --mattr=-sme2 - | FileCheck %s --check-prefix=CHECK-UNKNOWN
```
Comment (CarolineConcatto) [inline llvm/test/MC/AArch64/SME2p2/fmop4a-fp8-fp16-widening.s:8]: s/--mattr=-sme2/--mattr=-sme2p2

File: llvm/lib/Target/AArch64/SMEInstrFormats.td [hunk 2]
```diff
@@ -5126,3 +5126,38 @@ class sme2_luti4_vector_vg4_strided<bits<2> sz, bits<2> op, string mnemonic>
   let Inst{3-2}   = 0b00;
   let Inst{1-0}   = Zd{1-0};
 }
+
+class sme2_fp8_fp16_quarter_tile_outer_product<bit M, bit N, string mnemonic, RegisterOperand zn_ty, RegisterOperand zm_ty>
+    : I<(outs TileOp16:$ZAda),
+        (ins TileOp16:$_ZAda, zn_ty:$Zn, zm_ty:$Zm),
+        mnemonic, "\t$ZAda, $Zn, $Zm",
+         "", []>, Sched<[]> {
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/SMEInstrFormats.td:5134]: nit: along line 5134

File: llvm/lib/Target/AArch64/SMEInstrFormats.td [hunk 3]
```diff
@@ -5188,3 +5188,38 @@ class sme2_luti4_vector_vg4_strided<bits<2> sz, bits<2> op, string mnemonic>
...
+  let Inst{9} = N;
+  let Inst{8-6} = Zn;
+  let Inst{5-1} = 0b00100;
+  let Inst{0} = ZAda;
+
+  let Constraints = "$ZAda = $_ZAda";
+}
+
+multiclass sme2_fmop4a_fp8_fp16_2way<string mnemonic> {
+  // Single vectors
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/SMEInstrFormats.td:5214]: It is just a nit, but I think this is what I see usually when setting the bits:

  // Single vectors
  def _MZZ_BtoH : sme2_fp8_fp16_quarter_tile_outer_product<0b0, 0b0, mnemonic, ZPR8Mul2_Lo, ZPR8Mul2_Hi>;

  // Multiple and single vectors
  def _M2ZZ_BtoH : sme2_fp8_fp16_quarter_tile_outer_product<0b0, 0b1, mnemonic, ZZ_b_mul_r_Lo, ZPR8Mul2_Hi>;

  // Single and multiple vectors
  def _MZ2Z_BtoH : sme2_fp8_fp16_quarter_tile_outer_product<0b1, 0b0, mnemonic, ZPR8Mul2_Lo, ZZ_b_mul_r_Hi>;

  // Multiple vectors
  def _M2Z2Z_BtoH : sme2_fp8_fp16_quarter_tile_outer_product<0b1, 0b1, mnemonic, ZZ_b_mul_r_Lo, ZZ_b_mul_r_Hi>;

**Review by CarolineConcatto — APPROVED** [review]
I have a nit on how to represent bits, but otherwise LGTM

File: llvm/lib/Target/AArch64/SMEInstrFormats.td (hunk 3 above)
Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/SMEInstrFormats.td:5214]: I would like some argumentation about why do you think this is better?

Comment (CarolineConcatto) [reply llvm/lib/Target/AArch64/SMEInstrFormats.td:5214]: I was not saying it is better for 1 bit. I believe it does not make difference. 
But what I am suggesting is a pattern that I see most often (0b1/0b0) in this file and other related to instruction format.
In the llvm documentation it is also like you wrote for 1 bit:
https://llvm.org/docs/TableGen/ProgRef.html
I was just saying that this is not usually what I see in this file.

Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/SMEInstrFormats.td:5730]: Done.

**Pull request merged**